config/reader.go

```diff
@@ -49,29 +57,150 @@ func getAgentInfo() (*apppb.AgentInfo, error) {
+	// replaces all the . if they exist with _
+	return strings.ReplaceAll(version, ".", "_")
+}
+
+// there are 2 ways that we have configs, from rawfiles or from configs from the cloud
+// these methods do similar things, but in different orders
+
+func replacePlaceholders(bytes []byte, packageMap map[string]string) []byte {
+	// updates all the filepath placeholders with what we think they should be based on the packages
+	updatedBytes := placeholderRegexp.ReplaceAllFunc(bytes, func(b []byte) []byte {
+		updatedPath, ok := packageMap[verifyPlaceholder(string(b[:]))]
+		if !ok {
+			return b
+		}
+
+		return []byte(updatedPath)
+	})
+
+	return updatedBytes
 }
 
-func readFromCache(id string) (*Config, error) {
-	r, err := os.Open(getCloudCacheFilePath(id))
+func UpdatePlaceholdersInConfig(config *Config) (*Config, error) {
+	packages := config.Packages
+
+	bytes, err := config.MarshalJSON()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "cannot marshal config to bytes")
+	}
+	updatedBytes := replacePlaceholders(bytes, mapPlaceholderToSymlink(packages))
+
+	if err := config.UnmarshalJSON(updatedBytes); err != nil {
+		return nil, errors.Wrap(err, "cannot unmarshal bytes into config")
+	}
+	return config, nil
+}
+
+func mapPlaceholderToSymlink(packages []PackageConfig) map[string]string {
+
+	// stores the placeholders <> what they need to be replaced to
+	packageMap := make(map[string]string, len(packages))
+	for _, p := range packages {
+		packageMap[GetPackagePlaceholder(p)] = GenerateSymlinkDir(p)
 	}
+
+	return packageMap
+}
+
+func verifyPlaceholder(placeholder string) string {
+	// if its not a valid placeholder
+	if len(placeholder) < 3 && placeholder[0] != '$' && placeholder[1] != '{' && placeholder[len(placeholder)-1] != '}' {
+		return placeholder
+	}
+
+	// removes the ${} from the placeholder to match on it
+	cleaned := placeholder[2 : len(placeholder)-1]
+	return cleaned
+}
+
+func GenerateConfigFromFile(filepath string) (*Config, error) {
+	r, err := os.Open(filepath)
 	defer utils.UncheckedErrorFunc(r.Close)
 
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read config from file")
+	}
+
+	bytes, err := io.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	packages, err := GetPackagesFromFile(bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedBytes := replacePlaceholders(bytes, mapPlaceholderToSymlink(packages))
+
 	unprocessedConfig := &Config{
-		ConfigFilePath: "",
+		ConfigFilePath: filepath,
 	}
 
-	if err := json.NewDecoder(r).Decode(unprocessedConfig); err != nil {
-		// clear the cache if we cannot parse the file.
-		clearCache(id)
-		return nil, errors.Wrap(err, "cannot parse the cached config as json")
+	if err = unprocessedConfig.UnmarshalJSON(updatedBytes); err != nil {
+		return nil, errors.Wrap(err, "failed to decode config from json")
 	}
+
 	return unprocessedConfig, nil
+
+}
+
+// SymlinkPath is the one where the actual packages live with the correct versioning
```

Comment:
For this project, don't need to use the Symlinks, the config will point to the path in .data with the version, so would avoid the name and use `package filepath` instead.

The symlinks are created by package manager as convenience, pointing root/.viam/packages/myPackage to  root/.viam/modules/.data/orgId-myPackage

Reply:
in this case also how we do want to handle package filepaths, as their types are ml-model. Should it be 
`root/.viam/packages/ml_model.data/myPackage` or `root/.viam/ml_model/.data/myPackage`
(i.e. package prefix as well as ml_model). to me it makes sense to not have the package prefix if we aren't going to have that for modules but should be purely based on type.

Reply:
sounds good! -- just to confirm then what I am thinking, keep creating the symlink for modules but they actually will only be for backward compatibility and not actually be referenced anywhere. 

I think migrating to removing the symlinks should also be done in a separate PR.

Reply:
to keep backwards compatibility, for all packages already declared the placeholders will be the symlinks. For new packages, the placeholders resolve to the real roots.